lib/puppet/node.rb

```diff
@@ -113,15 +113,25 @@ def initialize(name, options = {})
   end
 
   # Merge the node facts with parameters from the node source.
-  def fact_merge
-    if @facts = Puppet::Node::Facts.indirection.find(name, :environment => environment)
-      @facts.sanitize
-      merge(@facts.values)
+  # @api public
+  # @param[optional, Puppet::Node::Facts] facts to merge into node parameters.
+  #   Will query Facts indirection if not supplied.
+  # @raise[Puppet::Error] Raise on failure to retrieve facts if not supplied
+  # @return[nil]
+  def fact_merge(facts = nil)
+    if facts.nil?
+      begin
+        @facts = Puppet::Node::Facts.indirection.find(name, :environment => environment)
+      rescue => detail
+        error = Puppet::Error.new(_("Could not retrieve facts for %{name}: %{detail}") % { name: name, detail: detail })
```

Comment:
you can supply the original error as the second argument to the constructor to save it as a wrapped exception